### PR TITLE
Support 20 bytes address

### PR DESCRIPTION
**Pull request by lxfind**
The current address is hard-coded to be 16 bytes. This works well for Diem, but could be insufficient for other systems. A common requirement is 20 bytes.
This PR adds a feature that gates the size of the address to be either 16 or 20 bytes, and change everywhere this is used to be using the right length.

Basically redo https://github.com/diem/diem/pull/10015 since the repo moved.

### CODE REVIEW

**Review by sblackshear — APPROVED** [review]
Accepting here since this was approved on the Diem side.

**Review by bmwill — DISMISSED** [review]

**Comment by mkurnikov** [comment]
Hi, in our Pontem Network fork of the Move compiler, we use 32 bytes as address length https://github.com/pontem-network/diem/blob/v1.5.1-pre/language/move-core/types/src/account_address.rs#L20. 
Is it possible to add even more configurability to this value?

**Comment by sblackshear** [comment]
> Hi, in our Pontem Network fork of the Move compiler, we use 32 bytes as address length https://github.com/pontem-network/diem/blob/v1.5.1-pre/language/move-core/types/src/account_address.rs#L20. Is it possible to add even more configurability to this value?

Yes, this is possible. There's some discussion about using const generics to allow arbitrary address lengths in https://github.com/diem/diem/pull/10015, but the conclusion is that this would be a significant refactoring.

To allow 32 byte addresses, the quickest path would be to send another PR mimicking this one that adds an `address32` feature. But would also be very happy (happier, in fact!) to support a PR performing the refactoring required to enable full configurability.

**Review by sblackshear — APPROVED** [review]

**Comment by lxfind** [comment]
@tnowacki There are unit test failures with this PR, but I am unable to reproduce it locally. All tests pass on my Mac. Not sure what's going on? Any insights on what might be happening here?

**Comment by tnowacki** [comment]
@lxfind 
1) This test is weird and duplicates a module for no reason
2) Let me check you your branch and take a look

File: language/move-core/types/Cargo.toml [hunk 1]
```diff
@@ -30,5 +30,6 @@ regex = "1.4.3"
 serde_json = "1.0.64"
 
 [features]
+address20 = []
```
Comment (tnowacki) [inline language/move-core/types/Cargo.toml:33]: Does this add the feature all the time? I'm confused

Comment (lxfind) [reply language/move-core/types/Cargo.toml:33]: No it just defines the feature. To enable it `address20` needs to be adde to `default` (or explicitly specified when adding the move-core-types dependency)

**Comment by tnowacki** [comment]
No idea whats up, rerunning the jobs to see if it was just flakiness? As in general flakiness or just that test. Or if that test is really broken somehow?

**Comment by tnowacki** [comment]
Same test again. Going to try running all tests on my mac and see if maybe it is some config poisoning thing or something between tests

**Comment by tnowacki** [comment]
I'm at a loss. Maybe @tzakian has some CLI related ideas?

**Comment by tnowacki** [comment]
Another possible thought, is your branch on your fork up to date? That is, are main on your fork and main on diem/move equivalent? 

I have seen some CI flakiness when those aren't in sync

**Comment by lxfind** [comment]
> Another possible thought, is your branch on your fork up to date? That is, are main on your fork and main on diem/move equivalent?
> 
> I have seen some CI flakiness when those aren't in sync

Yeah I rebased to main today.

**Comment by tnowacki** [comment]
Maybe @bmwill knows someone who can help debug this?

**Comment by tnowacki** [comment]
Only other thing I could imagine, just throwing debugging ideas out here, switch the address from 0x3 to something else for this test (maybe 0x1) just to rule out a bug with the address parsing code

**Comment by lxfind** [comment]
OK. So I reverted my commit and added some dummy changed (empty lines), and the test still failed. So pretty sure it's not related to my change.

**Comment by tnowacki** [comment]
As a stopgap measure, can you see how the test looks with renaming module A? Or changing the address? There is no reason whatsoever that the two modules need the same name between the two test files

**Comment by lxfind** [comment]
> As a stopgap measure, can you see how the test looks with renaming module A? Or changing the address? There is no reason whatsoever that the two modules need the same name between the two test files

Renaming it does fix the test failure. But it also kind of changed the semantics of the test (the test seems to be testing the duplicate module name on purpose?)

**Comment by lxfind** [comment]
I edited the test file so that it can pass.
Filed issue #22 to track the broken test.

**Review by tnowacki — APPROVED** [review]
